wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java

```diff
@@ -440,27 +467,11 @@ public Command getDefaultCommand(Subsystem subsystem) {
    * @param commands the commands to cancel
    */
   public void cancel(Command... commands) {
-    if (m_inRunLoop) {
-      m_toCancel.addAll(List.of(commands));
-      return;
-    }
-
     for (Command command : commands) {
-      if (command == null) {
-        DriverStation.reportWarning("Tried to cancel a null command", true);
-        continue;
-      }
-      if (!isScheduled(command)) {
-        continue;
-      }
-
-      m_scheduledCommands.remove(command);
-      m_requirements.keySet().removeAll(command.getRequirements());
-      command.end(true);
-      for (Consumer<Command> action : m_interruptActions) {
-        action.accept(command);
+      cancelCommand(command);
+      while (!m_toCancel.isEmpty() && !m_deferCancel) {
```

Comment:
Why is m_deferCancel checked?

Reply:
It prevents infinite recursion if a command is canceled from `end`; when the first command is canceled, `end` will get called, which calls `cancel` for the 2nd command. What happens is that `cancelCommand` gets called, and the command cancel gets deferred (we're already canceling the 1st command), but then the code gets stuck in while loop because it tries to cancel the command that was just deferred, but can't because we're in the middle of canceling a command.